Spectra: add stems layer to controller

SamWhited opened the pull request:
This PR adds a stems layer to the MIDI Fighter Spectra. As with the [Twister PR](https://github.com/mixxxdj/mixxx/pull/14931) this is for 2.6 so no particular rush.

Manual PR: mixxxdj/manual#776

SamWhited committed:
Spectra: add stems layer to controller

Signed-off-by: Sam Whited <sam@samwhited.com>

Swiftb0y reviewed (commented):
buncha nits. lgtm to me otherwise.

SamWhited commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -16,6 +16,73 @@ var MidiFighterSpectra;
+                    midi: [0x92, 0x5C + i],
+                    key: "enabled",
+                    type: components.Button.prototype.types.toggle,
+                    on: 66, // green
```
I'm familiar with `Object.freeze` but I'm not entirely clear on what you'd like frozen here? The value of "on"? I'm not sure how I'd even make that an object, or maybe I'm misunderstanding?

Swiftb0y replied:
I was suggesting to factor out the integer constants used in the color mapper (and here) into a constant dictionary in order to refer to it by name (instead of documenting the color in the adjacent comment).
Eg in this example you can then do 
```suggestion
                    on: COLORS.green,
```
(where `COLORS` is the frozen global object).

SamWhited replied:
Makes sense; done!

SamWhited commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -163,25 +213,29 @@ var MidiFighterSpectra;
                 output: function(_value, _group, control) {
                     for (let i = 0; i < 4; i++) {
-                        midi.sendShortMsg(this.midi[0], this.midi[1] + i, (control === this.midi[1] + i) ? this.on : this.off);
+                        for (const midino of this.midi[1]) {
+                            midi.sendShortMsg(this.midi[0], midino + i, (control === midino + i) ? this.on : this.off);
+                        }
```
I don't think I can override send for the output one because it needs the group from the XML file, I refactored these out into just the input and output functions though, so hopefully that keeps it a bit more clear? Happy to rework this again if I've missed something, hopefully this is more in line with the intent of the commonjs library though?

Swiftb0y replied:
Ah I see, I would then like to propose a slightly larger refactoring to the selectDeck button. You currently handle both physical buttons in the same component. This goes against the ComponentJS ethos (evident by how awkward it is to implement). If you instead bundle each button with its status LED into its own component, you can simplify the entire process by just overwriting `input` (to set the deck), `trigger` and `outGetValue`. Then just call `trigger` on all buttons in `activeDeck.setCurrentDeck`. Does that make sense? Is this reasonable?

SamWhited replied:
That makes sense to me, I was trying to keep the radio button logic contained in the button itself, but I'm not sure that really makes sense. setCurrentDeck seems like a sensible place to put it. I'll experiment with that now.

SamWhited replied:
I refactored this, but ran into an issue and it's not exactly what you asked for, but I think it works. I wasn't sure what `outGetValue` was for or how that would work, or what changes would need to be made to `trigger`. I also couldn't store a reference to the deck selection buttons on the deck itself, otherwise they would all be changed to a specific channel when you pressed one of them, so it resulted in a bit of a weird dependency loop where you can't call the first selection until after everything is constructed in init. LMK what you think.

SamWhited replied:
Hmm, didn't realize I had started testing with 2.5 by mistake. Switching back to 2.6 and now the stem buttons aren't getting their colors set properly. I'll take another look.

SamWhited replied:
Fixed, I had somehow broken the new channelFromStem function in ways that made absolutely no sense, I assume that somehow after testing and before committing I hit undo or pasted some stuff in the wrong place or something; anyways, fixed now. Everything is working again.

Swiftb0y commented on res/controllers/common-controller-scripts.js:
```diff
@@ -194,6 +194,15 @@ script.midiDebug = function(channel, control, value, status, group) {
+script.channelFromStem = function(stem) {
+    // Jank safety check that's faster than a regex and probably good enough.
+    if (stem.length !== 16 || stem.substring(0, 8) !== "[Channel" || stem.substring(9, 14) !== "Stem" || stem[15] !== "]") {
+        return "";
```
NIT: we usually use `undefined` for a `nullptr`-like value.

SamWhited replied:
Done, thanks!

SamWhited commented:
Opened a PR against the manual as well: mixxxdj/manual#776

Swiftb0y commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -224,45 +231,36 @@ var MidiFighterSpectra;
+                    input: function(_channel, control, value, _status, group) {
+                        MidiFighterSpectra.controller.activeDeck.setCurrentDeck(group);
+                    },
+                    setActive(group) {
```
Is this valid syntax in the ancient ES7 we're bound to by QJSEngine? It works in nodejs, but idk about here.

Swiftb0y replied:
[Apparently it is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#method_definitions). Lets stick to `property: function(...) {...}` for now, but feel free to use the shorthand notation in future mappings.

SamWhited replied:
Weird; I had no idea. Fixed for consistency either way!

Swiftb0y commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -224,45 +231,36 @@ var MidiFighterSpectra;
+                    input: function(_channel, control, value, _status, group) {
+                        MidiFighterSpectra.controller.activeDeck.setCurrentDeck(group);
+                    },
+                    setActive(group) {
+                        // Set the LED if this button represents the currently
+                        // selected deck.
+                        midi.sendShortMsg(this.midi[0], this.midi[1], (group === this.group) ? this.on : this.off);
+                    },
```
You'll want to un-inline (outline?) some functions here.

```js
{
    outGetValue: function() {
        return MidiFighterSpectra.controller.activeDeck.currentDeck === this.group
    }
    trigger: function() {
        this.output(this.outGetValue())); // this assumes a vanilla implementation of `output`, it probably doesn't work with your current impl OOTB
    }
}
```

Swiftb0y commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -162,11 +162,18 @@ var MidiFighterSpectra;
+
+        setCurrentDeck(newGroup) {
+            components.Deck.prototype.setCurrentDeck.call(this, newGroup);
+            for (const btn of MidiFighterSpectra.controller.selectDeck) {
+                btn.setActive(newGroup);
+            }
```
just call `.trigger()` on the buttons. Potentially even the entire deck.

SamWhited replied:
We can't call trigger here because these are doing two fundamentally different things; trigger triggers the output for the connection, ie. pulse the LED (or stop it pulsing) at `end_of_track`. However, the `setActive` stuff is changing the brightness or color of the LED depending on if the deck is active or not, which isn't something we can connect to. We don't want one function handling both because we'd be sending a lot of unnecessary extra MIDI signals (which I suppose isn't the end of the world, but still isn't ideal). Ie if we swapped decks we'd send a "stop blinking" or "start blinking" signal even though that is already happening when all we wanted to do was dim the LED.

Swiftb0y commented on res/controllers/DJ TechTools-MIDI Fighter Spectra-scripts.js:
```diff
@@ -224,45 +231,36 @@ var MidiFighterSpectra;
-                    if (engine.getSetting("pulseDeckSelect")) {
-                        for (let i = 0; i < 4; i++) {
-                            this.connections[i] = engine.makeConnection(`[Channel${i + 1}]`, "end_of_track", this.output.bind(this));
+            for (let i = 0; i < 8; i++) {
+                this.selectDeck[i] = new components.Button({
```
the button indicies 4-8 don't make much sense to me. Can you elaborate what their purpose is?

SamWhited replied:
There are 4 of these buttons across two layers, each group of 4 is a set of radio buttons where only one can be illuminated at a time (indicating which deck is selected). I just put them in a slice to have an easy way to reference them from the XML, I could equally name them "layer1Deck1Button" and "layer2Deck1Button" or something, but the loop made it easier to construct them.

SamWhited commented:
I'll go back and add some comements because it looks like there's some confusion about how the buttons are functioning here, but the fundamental problem is that they need to be connected to two different things: the `end_of_track` signal, and also a "deck change" signal. The second isn't actually a thing we connect to, so we can't just override `connect`, we have to implement the buttons two different illumination changes two different ways.

SamWhited committed:
Spectra: fix quick effect on/off setting name

Also make the default "off" color for quick effects the dim green
instead of completely off. This better fits with how the rest of the
buttons operate where "off" means "button not in use".

Signed-off-by: Sam Whited <sam@samwhited.com>

SamWhited commented:
Squashed this and then made some minor fixes to the color of the super button for the quick effects (to make it match the regular one on the decks).

SamWhited commented:
Gentle ping; does this look good? Not exactly what you were asking for per the comments above, but if I understand things correctly it can't be done that way for the reasons outlined. Thanks!

Swiftb0y reviewed (approved):
Yup. Good enough for me for now.
